lib/ts/recipe/webauthn/types.ts

```diff
@@ -0,0 +1,520 @@
+} from "../../ingredients/emaildelivery/types";
+import EmailDeliveryIngredient from "../../ingredients/emaildelivery";
+import { GeneralErrorResponse, NormalisedAppinfo, User, UserContext } from "../../types";
+import RecipeUserId from "../../recipeUserId";
+
+export type TypeNormalisedInput = {
+    relyingPartyId: TypeNormalisedInputRelyingPartyId;
+    relyingPartyName: TypeNormalisedInputRelyingPartyName;
+    getOrigin: TypeNormalisedInputGetOrigin;
+    getEmailDeliveryConfig: (
+        isInServerlessEnv: boolean
+    ) => EmailDeliveryTypeInputWithService<TypeWebauthnEmailDeliveryInput>;
+    override: {
+        functions: (
+            originalImplementation: RecipeInterface,
+            builder?: OverrideableBuilder<RecipeInterface>
+        ) => RecipeInterface;
+        apis: (originalImplementation: APIInterface, builder?: OverrideableBuilder<APIInterface>) => APIInterface;
+    };
+};
+
+export type TypeNormalisedInputRelyingPartyId = (input: {
+    request: BaseRequest | undefined;
+    userContext: UserContext;
+}) => string; // should return the domain of the origin
+
+export type TypeNormalisedInputRelyingPartyName = (input: {
+    request: BaseRequest | undefined;
+    userContext: UserContext;
+}) => string; // should return the app name
+
+export type TypeNormalisedInputGetOrigin = (input: { request: BaseRequest; userContext: UserContext }) => string; // should return the app name
+
+export type TypeInput = {
+    emailDelivery?: EmailDeliveryTypeInput<TypeWebauthnEmailDeliveryInput>;
+    relyingPartyId?: TypeInputRelyingPartyId;
+    relyingPartyName?: TypeInputRelyingPartyName;
+    getOrigin?: TypeInputGetOrigin;
+    override?: {
+        functions?: (
+            originalImplementation: RecipeInterface,
+            builder?: OverrideableBuilder<RecipeInterface>
+        ) => RecipeInterface;
+        apis?: (originalImplementation: APIInterface, builder?: OverrideableBuilder<APIInterface>) => APIInterface;
+    };
+};
+
+export type TypeInputRelyingPartyId =
+    | string
+    | ((input: { request: BaseRequest | undefined; userContext: UserContext }) => string);
+
+export type TypeInputRelyingPartyName =
+    | string
+    | ((input: { request: BaseRequest | undefined; userContext: UserContext }) => string);
+
+export type TypeInputGetOrigin = (input: { request: BaseRequest; userContext: UserContext }) => string;
+
+export type RecipeInterface = {
+    registerOptions(input: {
+        email: string;
+        relyingPartyId: string;
+        relyingPartyName: string;
+        origin: string;
+        timeout: number;
+        attestation: "none" | "indirect" | "direct" | "enterprise";
+        tenantId: string;
+        userContext: UserContext;
+    }): Promise<{
+        status: "OK";
+        webauthnGeneratedOptionsId: string;
+        rp: {
+            id: string;
+            name: string;
+        };
+        user: {
+            id: string;
+            name: string;
+            displayName: string;
```

Comment:
where does name and displayName come from? What effect does this have?

Reply:
it's the user email for both, as we don't have first/last name as inputs and we shouldn't disclose them either even if we would have them stored.

"The WebAuthn User Display Name (user.displayName) is a user-chosen, readable name for display in the app interface, not for authentication. It enhances user experience by providing a familiar, human-readable identifier, like a real name or nickname. " 

"The WebAuthn User Name (user.name) is a unique identifier within the RP’s system, crucial in scenarios where users might have similar Display Names. It’s used in the authentication process and can be an email, username, or any unique ID, ensuring each account is distinctively linked to its credentials. This name is vital for maintaining authentication integrity, especially when different accounts have similar Display Names. It can be any arbitrary string, offering flexibility in user identification."

See https://medium.com/@corbado_tech/webauthn-credential-id-user-name-user-id-user-handle-afe248a1d7c5#:~:text=The%20WebAuthn%20User%20Display%20Name,a%20real%20name%20or%20nickname.